File: LeetCode-Swift/LeetCode-Swift/动态规划/_198_打家劫舍.swift
//
//  _198_打家劫舍.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/11/25.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/house-robber/

import Foundation

class _198_打家劫舍 {
    func rob(_ nums: [Int]) -> Int {
        var first = 0
        var second = 0
        for num in nums {
            let tmp = second
            second = max(first + num, second)
            first = tmp
        }
        return max(first, second)
    }
}

/**
 输入: [1,2,3,1]
 输出: 4

 输入: [2,7,9,3,1]
 输出: 12
*/
func test_198() {
//    let nums = [1, 2, 3, 1]
    let nums = [2, 7, 9, 3, 1]
    let solution = _198_打家劫舍()
    print(solution.rob(nums))
}

File: LeetCode-C++/LeetCode-C++/动态规划/_198_打家劫舍.cpp
//
//  _198_打家劫舍.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/11/25.
//  Copyright © 2019 warden. All rights reserved.
//

#include "_198_打家劫舍.hpp"

int _198_打家劫舍::rob(vector<int>& nums) {
    int first = 0;  // 上上家
    int second = 0; // 上一家
    int tmp;
    for (int num : nums) {
        tmp = second;
        first += num;
        second = first > second ? first : second;
        first = tmp;
    }
    return first > second ? first : second;
}
